### PR TITLE
[Entities] Add properties setting automatic box rotation.

### PR DESCRIPTION
**Pull request by Andrey2470T**
This PR adds two new object properties named as `synchronize_cbox_rotation_with_dir` and `synchronize_sbox_rotation_with_dir`, first of which aims with collision box, second with selection box. If any of them is set as true, the rotation of the corresponding box will follow the entitiy`s rotation. It may rotate _only under the right angle_ that is calculated based on the nearest value of pitch, yaw, roll that is multiple of 90 degrees. 

## To do

- Fix a bug when the boxes get skew on rotation around X/Z axises in some cases.

## How to test
1. Go to the dev test game.
2. Call `/spawnentity testentity:3d_dungeon_master`.
3. Use the arrows adjusting rotation around specific axises (right-click to rotate, next right-click to stop).

### CODE REVIEW

File: src/server/unit_sao.cpp [hunk 1]
```diff
@@ -38,6 +38,91 @@ ServerActiveObject *UnitSAO::getParent() const
...
+}
+
+void UnitSAO::setBoxRotation(v3f rotation, const std::string &box_type)
+{
+	ObjectProperties *props = accessObjectProperties();
+
+	aabb3f *box = box_type == "collision"
+					? &props->collisionbox
+					: (box_type == "selection" ? &props->selectionbox 
+					   				: nullptr);
```
Comment (Andrey2470T) [inline src/server/unit_sao.cpp:68]: How many tabs is it necessary to do here?

**Comment by lhofhansl** [comment]
This is clever. I was wondering how we can improve entity boxes when the entities rotate. Of course arbitrary rotation makes collision detected too expensive, but in multiples on 90 degrees this can definitely work.

File: src/server/unit_sao.cpp (hunk 1 above)
Comment (lhofhansl) [reply src/server/unit_sao.cpp:68]: I think you pull in the first two line line by one tab, and third by two tabs.

But perhaps this is nicer as an if-statement?
```
aabb3f *box = nullptr;
if (box_type == "collision)
        box = &props->collisionbox;
else if (box_type == "selection")
        box = &props->selectionbox;
```

But either is fine.

**Comment by appgurueu** [comment]
I'd prefer the box be rotated and the maximum axis-aligned bounding box of the resulting box be used.

**Comment by Andrey2470T** [comment]
> I'd prefer the box be rotated and the maximum axis-aligned bounding box of the resulting box be used.

The boxes have become to rotate due to this PR and what do you mean by "maximum axis-aligned bounding box of the resulting box be used"?

**Comment by lhofhansl** [comment]
> Fix a bug when the boxes get skew on rotation around X/Z axises in some cases.

Is this still an issue?

> I'd prefer the box be rotated and the maximum axis-aligned bounding box of the resulting box be used.

@appgurueu I'm afraid I don't follow. Could you explain a bit more or rephrase?

**Comment by Andrey2470T** [comment]
> > Fix a bug when the boxes get skew on rotation around X/Z axises in some cases.
> 
> Is this still an issue?
> 
Yes. And I don't know how to fix that yet. The problem is boxes rotate oppositely the direction where they should do around X and Z. Maybe incorrect sign of rotation, but I didn't notice in the code anything like.

**Comment by appgurueu** [comment]
> > I'd prefer the box be rotated and the maximum axis-aligned bounding box of the resulting box be used.
> 
> The boxes have become to rotate due to this PR and what do you mean by "maximum axis-aligned bounding box of the resulting box be used"?

If you rotate a axis-aligned box, you (might) get a non-axis-aligned box. That's impractical for collision checking. So, take the bounding box of this rotated box and use it as an axis-aligned box.

**Comment by Andrey2470T** [comment]
> > > I'd prefer the box be rotated and the maximum axis-aligned bounding box of the resulting box be used.
> > 
> > 
> > The boxes have become to rotate due to this PR and what do you mean by "maximum axis-aligned bounding box of the resulting box be used"?
> 
> If you rotate an axis-aligned box, you (might) get a non-axis-aligned box. That's impractical for collision checking. So, take the bounding box of this rotated box and use it as an axis-aligned box.

I'm sure that will never happen. Because internally this PR only sets the corresponding object properties, it doesn't change the behaviour of the boxes themselves when they rotate. Maximally what can happen is distorting of boxes sizes if rotation counting off +Z axis is not multiple of 90 degrees. Even regardless that, the boxes will stay axis-aligned.

**Comment by TheTermos** [comment]
> I was wondering how we can improve entity boxes when the entities rotate.

if an entity can rotate around Y the box' base must be square.
if it can rotate in 3d, the box must be a cube.

If I understand this PR correctly, it will lead to many glitches. If a malformed box is touching nodes and then its dimensions are swapped, then it will find itself partially inside nodes, and the entity will be able to glitch through a solid layer of nodes.

I suggest closing this asap and using the guidelines above instead. No new bugs and the code doesn't swell another 100+ redundant lines, a win-win situation in my book.

**Comment by appgurueu** [comment]
> > I was wondering how we can improve entity boxes when the entities rotate.
> 
> if an entity can rotate around Y the box' base must be square.
> if it can rotate in 3d, the box must be a cube.
> 
> If I understand this PR correctly, it will lead to many glitches. If a malformed box is touching nodes and then its dimensions are swapped, then it will find itself partially inside nodes, and the entity will be able to glitch through a solid layer of nodes.
> 
> I suggest closing this asap and using the guidelines above instead. No new bugs and the code doesn't swell another 100+ redundant lines, a win-win situation in my book.

I agree. We can't solve this with a hack. A proper implementation would have to make use of rotational velocity and then check if any collisions happen during rotation, as to prevent such situations. The current collision code is quite robust and accordingly probably handles this well, but still.

A starting point might be having selectionboxes rotate. Rotating boxes per step is already possible from Lua.

**Comment by Andrey2470T** [comment]
> > > I was wondering how we can improve entity boxes when the entities rotate.
> > 
> > 
> > if an entity can rotate around Y the box' base must be square.
> > if it can rotate in 3d, the box must be a cube.
> > If I understand this PR correctly, it will lead to many glitches. If a malformed box is touching nodes and then its dimensions are swapped, then it will find itself partially inside nodes, and the entity will be able to glitch through a solid layer of nodes.
> > I suggest closing this asap and using the guidelines above instead. No new bugs and the code doesn't swell another 100+ redundant lines, a win-win situation in my book.
> 
> I agree. We can't solve this with a hack. A proper implementation would have to make use of rotational velocity and then check if any collisions happen during rotation, as to prevent such situations. The current collision code is quite robust and accordingly probably handles this well, but still.
> 
I would think this glitch described above would rather to cause by _bad_ collision code itself, than the implementation of my PR. In my view the ideal collision code would not allow to go entities through node layers in similar situations, it would attempt to "pull out" them from there. 

> A starting point might be having selectionboxes rotate. Rotating boxes per step is already possible from Lua.
I just wanted to give generalized API for modders could obtain that feature by only enabling the options without a need to write the code by themselves and to avoid as a bad thing as "inventing bikes".

**Comment by TheTermos** [comment]
> In my view the ideal collision code would not allow to go entities through node layers in similar situations, it would attempt to "pull out" them from there.

Nope, 'pulling out' introduces yet another category of nasty bugs.

If a mod implements a rotating mob with a collision box which bottom face isn't a square, that should be considered the mod's error, so this PR is not only harmful, but also pointless. Please close.

**Comment by Andrey2470T** [comment]
> Nope, 'pulling out' introduces yet another category of nasty bugs.
> 
Why?
> If a mod implements a rotating mob with a collision box which bottom face isn't a square, that should be considered the mod's error...
>
I just wonder to hear that. There are tons of mods that define collision/selection boxes of entities whose bases are non-square. Also, lua_api.txt doesn't report about that anything.

I intend to really close this PR due to that bug and keep closed until that pulling out is not released.

Also, it's worth to mention that feature is already particularly done for dropped itemstacks. When they are inside a collision box, that box pulls out them outside. So, it can argue it is not a "category of nasty bugs" as you said above.

**Comment by TheTermos** [comment]
If you have moving and shapeshifting boxes, by the time boxes get to intersect there's no guarantee there's room for them to be pulled out.
If you can remember, before 5.2.0 there used to be more collision detection bugs, some of them caused by pulling out .

> There are tons of mods that define collision/selection boxes of entities whose bases are non-square. Also, lua_api.txt doesn't report about that anything.

lua_api.txt doesn't seem to concern itself with any guidelines regarding collision boxes whatsoever, anyway this stuff becomes pretty obvious once you get to know how they work. E.g. if you want an entity to appear as if it stands on the ground the boxes bottom must be aligned with the model's bottom - no need to bother the documentation with it.

Some modders don't understand how collision boxes work, and why they must be axis aligned, so they fail to take rotation into account and naively assume that if their model is of different length than width, so should be the collision box.

**Comment by lhofhansl** [comment]
Maybe I misunderstood. If you have an object that much longer than wide and high (say), we would not want a cube collision box.  I get it's hard to get right, but that does not change that you do not want that. Of course modders expect their collision to rotate with the entity, that's the natural assumption.

In that case aligning the collision along 90 degree angles is better than not doing that.

Of course we cannot capture the rotation in between. So what? Right now you either have giant cubes that lead to unnatural behavior or you have small ones and entities partially scale through walls, or you have boxes that do not align with the visual shape of the entity at all.

We're doing the "Minetest thing" again. Someone has a solution, it's not prefect so we're not doing it, meanwhile nobody proposed the perfect solution because it's too hard to do. And so... Nothing happens. Again.

If simplicity is really that important, let's put our money were our mouth is, and make it a cube, always, (or perhaps a sphere is even easier, just a radius, and simple distance calculations) and then remove all the collision code that does more complicated things. I'm sure that would remove a lot of glitches.

**Comment by appgurueu** [comment]
> If simplicity is really that important, let's put our money were our mouth is, and make it a cube, always, (or perhaps a sphere is even easier, just a radius, and simple distance calculations) and then remove all the collision code that does more complicated things. I'm sure that would remove a lot of glitches.

First of all, modders can already set the collisionbox on_step. The engine does not have to be doing this for them. "Unclean" implementations should be left to modders as a possibility (as they currently are), and not be endorsed as engine features. This would introduce *horrible* glitching.

If we want a proper way to do this, we'd need rotational velocity / acceleration, and have to simulate the rotation of the box - and where it would be stopped.

**Comment by TheTermos** [comment]
> In that case aligning the collision along 90 degree angles is better than not doing that.

Not really, it still doesn't help with misaligned boxes most of the time and introduces problems far worse than misaligned boxes.

Collision detection more sophisticated than aaboxes is fairly complex stuff, most AAA games don't go there so let's be real. If I'm dealing with  long objects, let's take a spear for example, I put a small cubic box where it matters the most - near the tip. Nobody's gonna mind lack of collision with the shaft.

**Pull request closed without merging**